ros/angel_system_nodes/angel_system_nodes/uho_activity_detector.py

```diff
@@ -0,0 +1,336 @@
+
+        if len(self._frames) >= self._frames_per_det:
+            frame_stamp_set = self._frame_stamps[:self._frames_per_det]
+            ready_to_predict = False
+
+            # If the source stamp for this detection message is after or equal to
+            # the last frame stamp in the current set, then we have all of the
+            # detections and can move onto processing this set of frames.
+            msg_nsec = msg.source_stamp.sec * 10e9 + msg.source_stamp.nanosec
+            frm_nsec = frame_stamp_set[-1].sec * 10e9 + frame_stamp_set[-1].nanosec
+            if msg_nsec >= frm_nsec:
+                ready_to_predict = True
+
+            # Need to wait until the object detector has processed all of these frames
+            if not ready_to_predict:
+                log.info(f"Waiting for more object detection results")
+                return
+
+            frame_set = self._frames[:self._frames_per_det]
+            lhand_pose_set = self._hand_poses['lhand'][:self._frames_per_det]
+            rhand_pose_set = self._hand_poses['rhand'][:self._frames_per_det]
+
+            aux_data = dict(
+                lhand=lhand_pose_set,
+                rhand=rhand_pose_set,
+                dets=[],
+                bbox=[],
+            )
+
+            # Format the object detections into descriptors and bboxes
+            idxs_to_remove = []
+            for idx, det in enumerate(self._obj_dets):
+                #print(det.source_stamp, frame_stamp_set[0], frame_stamp_set[-1])
+                if det.num_detections == 0:
+                    log.info(f"no dets, det source: {det.source_stamp}")
+                    continue
+
+                det_source_stamp_nsec = det.source_stamp.sec * 10e9 + det.source_stamp.nanosec
+                first_frm_nsec = frame_stamp_set[0].sec * 10e9 + frame_stamp_set[0].nanosec
+                last_frm_nsec = frame_stamp_set[-1].sec * 10e9 + frame_stamp_set[-1].nanosec
+
+                # Check that this detection is within the range of time
+                # for the current frame set
+                if det_source_stamp_nsec < first_frm_nsec:
+                    # Detection is before the first frame in this set,
+                    # so we can remove it
+                    idxs_to_remove.append(idx)
+                    continue
+                elif det_source_stamp_nsec > last_frm_nsec:
+                    # Detection is after the last frame in this set,
+                    # so keep it for later
+                    continue
+
+                # Get the topk detection confidences
+                det_confidences = (
+                    torch.Tensor(det.label_confidences)
+                    .reshape((det.num_detections, len(det.label_vec)))
+                )
+                topk, i = torch.topk(det_confidences.flatten(), self._topk)
+
+                ind_2d = np.array(np.unravel_index(i.numpy(), det_confidences.shape)).T
+                top_det_idx = ind_2d[:, 0]
```

Comment:
I'm not sure if this is what we want. Since this is flattening *all* confidences, it can be the case where the top confidences are in the same descriptor (all descriptors "even", one has two spikes).

The following looks as descriptor max confidence values specifically before going into the topk function, should avoid jumping out of torch format (maybe the tiniest amount more efficient), and more readable IMO.
```suggestion
                det_max_confidences = det_confidences.max(axis=1).values
                _, top_det_idx = torch.topk(det_max_confidences, self._topk)
```